lib/auth/auth_with_roles_test.go

```diff
@@ -955,6 +957,176 @@ func TestReplaceRemoteLocksRBAC(t *testing.T) {
 	}
 }
 
+// TestIsMFARequiredMFADB tests isMFARequest logic per database protocol where different role matchers are used.
+func TestIsMFARequiredMFADB(t *testing.T) {
+	const (
+		databaseName = "test-database"
+		userName     = "test-username"
+	)
+
+	type modifyRoleFunc func(role types.Role)
+	tests := []struct {
+		name               string
+		userRoleRequireMFA bool
+		checkMFA           require.BoolAssertionFunc
+		modifyRoleFunc     modifyRoleFunc
+		dbProtocol         string
+		req                *proto.IsMFARequiredRequest
+	}{
+		{
+			name:       "RequireSessionMFA enabled MySQL protocol doesn't match database name",
+			dbProtocol: libdefaults.ProtocolMySQL,
+			req: &proto.IsMFARequiredRequest{
+				Target: &proto.IsMFARequiredRequest_Database{
+					Database: &proto.RouteToDatabase{
+						ServiceName: databaseName,
+						Protocol:    libdefaults.ProtocolMySQL,
+						Username:    userName,
+						Database:    "example",
+					},
+				},
+			},
+			modifyRoleFunc: func(role types.Role) {
+				roleOpt := role.GetOptions()
+				roleOpt.RequireSessionMFA = true
+				role.SetOptions(roleOpt)
+
+				role.SetDatabaseUsers(services.Allow, []string{types.Wildcard})
+				role.SetDatabaseLabels(types.Allow, types.Labels{types.Wildcard: {types.Wildcard}})
+				role.SetDatabaseNames(types.Allow, nil)
+			},
+			checkMFA: require.True,
+		},
+		{
+			name:       "RequireSessionMFA disabled",
+			dbProtocol: libdefaults.ProtocolMySQL,
+			req: &proto.IsMFARequiredRequest{
+				Target: &proto.IsMFARequiredRequest_Database{
+					Database: &proto.RouteToDatabase{
+						ServiceName: databaseName,
+						Protocol:    libdefaults.ProtocolMySQL,
+						Username:    userName,
+						Database:    "example",
+					},
+				},
+			},
+			modifyRoleFunc: func(role types.Role) {
+				roleOpt := role.GetOptions()
+				roleOpt.RequireSessionMFA = false
+				role.SetOptions(roleOpt)
+
+				role.SetDatabaseUsers(services.Allow, []string{types.Wildcard})
+				role.SetDatabaseLabels(types.Allow, types.Labels{types.Wildcard: {types.Wildcard}})
+				role.SetDatabaseNames(types.Allow, nil)
+			},
+			checkMFA: require.False,
+		},
+		{
+			name:       "RequireSessionMFA enabled Postgres protocol database name doesn't match",
+			dbProtocol: libdefaults.ProtocolPostgres,
+			req: &proto.IsMFARequiredRequest{
+				Target: &proto.IsMFARequiredRequest_Database{
+					Database: &proto.RouteToDatabase{
+						ServiceName: databaseName,
+						Protocol:    libdefaults.ProtocolPostgres,
+						Username:    userName,
+						Database:    "example",
+					},
+				},
+			},
+			modifyRoleFunc: func(role types.Role) {
+				roleOpt := role.GetOptions()
+				roleOpt.RequireSessionMFA = true
+				role.SetOptions(roleOpt)
+
+				role.SetDatabaseUsers(services.Allow, []string{types.Wildcard})
```

Comment:
```suggestion
				role.SetDatabaseUsers(types.Allow, []string{types.Wildcard})
```

Reply:
`services.Allow` is just an alias for `types.Allow` which should have been removed. Fixed in this PR - https://github.com/gravitational/teleport/pull/8441